### DISCUSSION TITLE
Building 86Box in macOS

### DISCUSSION THREAD

**Discussion by almeath** [86Box/86Box#2140] [Q&A] [answered]
I saw there were some updates to the codebase in recent months to enable a stable Linux build, so I decided to try building in macOS 10.14.6 (Mojave).

This is what I have done so far:

1. Manually install rtmidi with cmake, because for some reason the brew install repeatedly fails on file verification (perhaps because I am using an "unsupported" version of macOS).
2. Installed qt5 via brew (it installs to /usr/local/Cellar/qt@5)
3. Ran the following commands in terminal:

```
cmake_policy() CMP0079 NEW

export CMAKE_PREFIX_PATH=/usr/local/Cellar/qt@5/5.15.2_1/ 

cmake -S <local path to source> -B <local path to build folder>

```
Compilation then (apparently) succeeds, as there are no build errors. The output was as follows:

```
192-168-1-100% cmake -S /Users/almeath/Desktop/86Box-master -B /Users/almeath/Desktop/build 
-- The C compiler identification is AppleClang 11.0.0.11000033
-- The CXX compiler identification is AppleClang 11.0.0.11000033
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Freetype: /usr/local/lib/libfreetype.dylib (found version "2.11.1") 
-- Found OpenAL: /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenAL.framework  
-- Found ZLIB: /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/lib/libz.tbd (found version "1.2.11") 
-- Found PNG: /usr/X11R6/lib/libpng.dylib (found version "1.6.37") 
-- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.2") 
-- Checking for module 'rtmidi'
--   Found rtmidi, version 5.0.0
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/almeath/Desktop/build
```

I end up with a src folder with the following contents:

<img width="561" alt="Screen Shot 2022-02-19 at 4 32 52 pm" src="https://user-images.githubusercontent.com/5726214/154793414-b65d3dcc-dd07-47d0-bb9c-9ded5c9666cd.png">

The resulting '86box.app' is a 1k file with no resources or executable contents, so I know I am doing something wrong but I am not sure what. Any suggestions would be appreciated.

**Comment by dhrdlicka** [comment, marked as answer]
You haven't actually built the project, you only generated the build scripts.

The [GitHub action](https://github.com/86Box/86Box/blob/master/.github/workflows/cmake.yml) for Mac build goes roughly like this (ran from the 86Box source directory):
```console
$ cmake -S . -B build --toolchain cmake/flags-gcc-x86_64.cmake --preset regular -D QT=ON -D Qt5_DIR=/usr/local/Cellar/qt@5/5.15.2_1/lib/cmake/Qt5 -D Qt5LinguistTools_DIR=/usr/local/Cellar/qt@5/5.15.2_1/lib/cmake/Qt5LinguistTools/
$ cmake --build build
$ cmake --install build --prefix ./build/artifacts
```

You should then end up with a 86Box.app bundle in `builds/artifacts`.

**Comment by dob205** [comment]
And with the latest changes everything should be fine regarding the compilation as a complete bundle.

**Comment by almeath** [comment]
Thanks for these suggestions. Almost there, in that an 86box app is built, but I am trying to overcome a cmake error right at the end of the --install command:

```
CMake Error at /usr/local/Cellar/cmake/3.22.2/share/cmake/Modules/BundleUtilities.cmake:909 (message):
  Command failed:

   '/usr/bin/install_name_tool' '-change' '/usr/local/Cellar/qt@5/5.15.2_1/lib/QtCore.framework/Versions/5/QtCore' '@executable_path/../Frameworks/QtCore.framework/Versions/5/QtCore' '-change' '/usr/local/Cellar/qt@5/5.15.2_1/lib/QtGui.framework/Versions/5/QtGui' '@executable_path/../Frameworks/QtGui.framework/Versions/5/QtGui' '-change' '/usr/local/Cellar/qt@5/5.15.2_1/lib/QtWidgets.framework/Versions/5/QtWidgets' '@executable_path/../Frameworks/QtWidgets.framework/Versions/5/QtWidgets' '-delete_rpath' '@executable_path/../Frameworks' '-delete_rpath' '@loader_path/Frameworks' '-id' '@executable_path/../Frameworks/QtPrintSupport.framework/Versions/5/QtPrintSupport' '/Users/almeath/Desktop/86Box-master/build/artifacts/86Box.app/Contents/Frameworks/QtPrintSupport.framework/Versions/5/QtPrintSupport'
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.22.2/share/cmake/Modules/BundleUtilities.cmake:984 (fixup_bundle_item)
  build/src/qt/cmake_install.cmake:64 (fixup_bundle)
  build/src/cmake_install.cmake:131 (include)
  build/cmake_install.cmake:42 (include)
```

The app crashes when launched, and in the terminal the output is:

```
192-168-1-100% /Users/almeath/Desktop/86Box-master/build/artifacts/86Box.app/Contents/MacOS/86Box 
objc[67225]: Class QMacAutoReleasePoolTracker is implemented in both /usr/local/Cellar/qt@5/5.15.2_1/lib/QtCore.framework/Versions/5/QtCore (0x10977eb00) and /Users/almeath/Desktop/86Box-master/build/artifacts/86Box.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x10854cb00). One of the two will be used. Which one is undefined.
objc[67225]: Class QT_ROOT_LEVEL_POOL__THESE_OBJECTS_WILL_BE_RELEASED_WHEN_QAPP_GOES_OUT_OF_SCOPE is implemented in both /usr/local/Cellar/qt@5/5.15.2_1/lib/QtCore.framework/Versions/5/QtCore (0x10977eb78) and /Users/almeath/Desktop/86Box-master/build/artifacts/86Box.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x10854cb78). One of the two will be used. Which one is undefined.
objc[67225]: Class KeyValueObserver is implemented in both /usr/local/Cellar/qt@5/5.15.2_1/lib/QtCore.framework/Versions/5/QtCore (0x10977eba0) and /Users/almeath/Desktop/86Box-master/build/artifacts/86Box.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x10854cba0). One of the two will be used. Which one is undefined.
objc[67225]: Class RunLoopModeTracker is implemented in both /usr/local/Cellar/qt@5/5.15.2_1/lib/QtCore.framework/Versions/5/QtCore (0x10977ebf0) and /Users/almeath/Desktop/86Box-master/build/artifacts/86Box.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore (0x10854cbf0). One of the two will be used. Which one is undefined.
QObject::moveToThread: Current thread (0x7fdad6e0ae00) is not the object's thread (0x7fdab6c0a3e0).
Cannot move to target thread (0x7fdad6e0ae00)

You might be loading two sets of Qt binaries into the same process. Check that all plugins are compiled against the right Qt binaries. Export DYLD_PRINT_LIBRARIES=1 and check that only one set of binaries are being loaded.
qt.qpa.plugin: Could not load the Qt platform plugin "cocoa" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: cocoa.

zsh: abort   

```

The install_name_tool failed, so it is failing to resolve a conflict between the QtCore.framework in the app bundle and the QT package installed on my Mac.

**Reply by dob205** [reply]
Try to delete the build folder and pull the newest changes that includes a fix for that (did that with a M1 Pro Mac).

**Comment by almeath** [comment]
So far, I am getting the same error with fresh source code from both master and your own branch, @dob205

I will see if I can re-produce these results on another Mac, to rule out any issues specific to my build system.

**Comment by almeath** [comment]
I was able to compile under macOS Monterey (12.1), so we could potentially put this down to an incompatibility with Homebrew under Mojave, as they do post warnings about that every time you run 'brew update'. When launched, the 86Box app displays the expected error message about not finding any ROMs.

There seems to be no logical reason why the resulting app should not at least run under Mojave. However, it does not, as it presents an error stating "86Box cannot be opened because of a problem". Apple provides very informative error messages these days.

I launched the binary from within the app package, and was presented with the following error in the terminal:

```
192-168-1-100% /Users/almeath/Desktop/86Box.app/Contents/MacOS/86Box 
dyld: Library not loaded: @executable_path/../Frameworks/libpng16.16.dylib
  Referenced from: /Users/almeath/Desktop/86Box.app/Contents/MacOS/86Box
  Reason: no suitable image found.  Did find:
	/Users/almeath/Desktop/86Box.app/Contents/MacOS/../Frameworks/libpng16.16.dylib: cannot load 'libpng16.16.dylib' (load command 0x80000034 is unknown)
	/Users/almeath/Desktop/86Box.app/Contents/MacOS/../Frameworks/libpng16.16.dylib: cannot load 'libpng16.16.dylib' (load command 0x80000034 is unknown)
zsh: abort      /Users/almeath/Desktop/86Box.app/Contents/MacOS/86Box
```

The thing is, libpng16.16.dylib is sitting there under Contents/Frameworks within the app bundle, and it is also present on my system under /usr/local/Cellar/libpng/1.6.37

It seems I cannot install (or run) the later version of libpng 16.16 under Mojave, so that will make 86Box only compatible with Catalina (10.15) and higher.

**Reply by dob205** [reply]
It might be due to homebrew being a bit fickle, in theory macports should work better for that case. If you want to assist with development, feel free to join our Discord or open a PR.

**Comment by almeath** [comment]
I noticed that my install of libpng 1.6.37 had a copy of "libpng16.16.dylib" under the lib sub-folder, dated 8 September 2019. I dragged this dylib into the Contents/Frameworks folder of the 86Box app, replacing the version that had been installed during the build process. 

That solved the problem, as now the app will launch. It may or may not be transportable to another system, but it appears to be a successful work-around to get it working on my Mojave system.
